### ISSUE TITLE
SPEEDY shortwave heating is applied on only one step in three (nstrad gate drops the tendency)

### ISSUE THREAD

**Issue by dhruvbalwada** [climate-analytics-lab/jax-gcm#752]
**Affects:** every release from v1.0.0 through v2.0.1 and the current `main`/`dev` branches. The v1.1.1 tag also carries it, so the configuration described in the GMD paper may have been affected, depending on which commit produced its runs.
**Introduced by:** PR #151 "Fix shortwave timestepping" (merged 2025-05-21, commit `91855031`).
**Symptom:** the atmosphere receives roughly one third of the shortwave energy it absorbs. Globally ~35-40 W/m² of column heating is missing; the TOA-minus-surface energy budget does not close (39 W/m² in a 1-year T31 run); the atmosphere is ~5 K too cold in the mean.

## 1. What the problem is

SPEEDY evaluates shortwave radiation only every `nstrad` = 3 time steps (an optimisation, `physics.f90`), but it *keeps* the resulting heating rate (`tt_rsw`) and adds it to the temperature tendency on every step. JCM implements the first half of that pattern and not the second.

In `jcm/physics/radiation/speedy_shortwave.py`, `get_shortwave_rad_fluxes` computes the fluxes on every call and then does

```python
return jax.lax.cond(physics_data.shortwave_rad.compute_shortwave,
                    lambda: tendencies, lambda: zero_tendencies), physics_data
```

`compute_shortwave` is true on one step in three (`set_physics_flags` in `jcm/physics/speedy/speedy_terms.py`). On the other two steps the shortwave temperature tendency is **zero**; nothing caches or re-applies the last heating rate. The longwave routine (`speedy_longwave.py`) builds its tendency from the longwave absorption only, so it does not compensate.

Because the fluxes and diagnostics (`ftop`, `rsds`, `rsns`, `dfabs`) are still recomputed and written out on every step, the output looks self-consistent: the surface receives its full shortwave, the TOA diagnostic reports the full absorption, but the air is only heated one step in three. That is also why the gate as written saves no compute time.

The three ECHAM-style radiation backends in `jcm/physics/radiation/__init__.py` already implement the correct pattern (`cached_radiation_tendency` re-emits cached heating rates between radiation steps); the SPEEDY shortwave term was never wired to a cache.

Consequences beyond the budget: the missing heating is concentrated in the tropical troposphere (water vapour, clouds) and the stratosphere (ozone), so the meridional heating gradient, static stability, hydrological cycle and the jets are all affected, and implied heat transports from the energy budget are invalid.

## 2. Proposed fix (keeps the intended speed-up)

A pull request with the fix against `dev` follows. Three parts:

1. **Cache.** Add a `heating_rate` field to `SWRadiationData` (`physics_data.py`), the shortwave carry slot, holding the last computed temperature tendency (K/s). Documented in `units_table.csv`.
2. **Compute-and-store / replay.** In `get_shortwave_rad_fluxes`, put the whole flux computation inside the `lax.cond`: on a radiation step compute the fluxes, store the tendency in `heating_rate`, and return it; on the other steps return the cached tendency and the unchanged carry.
3. **Skip clouds too.** `get_clouds` is likewise only evaluated on radiation steps (it produces no tendency; its fields are carried).

```python
def _compute(_):
    _, new_pd, _, _, _, tend = shortwave_rad_fluxes((state, physics_data, parameters, forcing, terrain, zero))
    sw = new_pd.shortwave_rad.copy(heating_rate=tend.temperature)
    return tend, new_pd.copy(shortwave_rad=sw)

def _replay(_):
    return PhysicsTendency.zeros(shape=shape, temperature=physics_data.shortwave_rad.heating_rate), physics_data

return jax.lax.cond(physics_data.shortwave_rad.compute_shortwave, _compute, _replay, None)
```

No zenith-angle rescaling is needed on replay because SPEEDY (and JCM) use daily-mean insolation with no diurnal cycle. The radiative diagnostics are carried unchanged between radiation steps, which is consistent with the heating actually applied and with what Fortran SPEEDY writes.

Existing tests pass (`speedy_shortwave_test.py`, `speedy_longwave_test.py`, `speedy_terms_test.py` on `dev`).

### 3a. Column energy closure (JCM alone, T31, default aquaplanet forcing, 1 year, last 180 days)

TOA net = `shortwave_rad.ftop - longwave_rad.ftop`; surface net = `surface_flux.hfluxn.1`; column energy tendency from cp·T + L·q integrated in pressure. Residual = (TOA - surface) - dE/dt should be ~0 (SPEEDY-class physics is not exactly conservative; ~2 W/m² is expected from missing frictional heating, the sponge layer and the longwave corrections).

| configuration | TOA − surface (W/m²) | dE/dt (W/m²) | residual (W/m²) | mean atm. T (K) |
|---|---|---|---|---|
| shipped, nstrad = 3 | 39.1 | −0.1 | **39.2** | 237.1 |
| shipped, nstrad = 1 (every step) | 2.6 | 0.1 | 2.5 | 242.5 |
| **fixed, nstrad = 3** | 2.8 | −0.1 | **2.9** | 242.5 |
| fixed, nstrad = 1 | 2.6 | 0.1 | 2.5 | 242.6 |

The fixed code with the 3-step gate reproduces the every-step reference; the shipped code loses 37 W/m² and runs 5 K colder.

![column imbalance, SW absorption and temperature profile, before and after](https://raw.githubusercontent.com/dhruvbalwada/jax-gcm/docs/shortwave-issue/docs/issues/shortwave-cache/fig_closure_before_after.png)

Zonal structure of TOA − surface in the shipped code (W/m², 1-year run): +112 at 13°S, +89 at 2°N, −25 at 58°S, −31 at 61°N. With the fix the pattern is the physical convergence of atmospheric heat transport with a global mean near zero.

### 3b. Cost per step (T31, GPU V100, warm, 120-day run)

| configuration | ms per simulated day |
|---|---|
| shipped, nstrad = 3 (fluxes computed every step, heating dropped) | 105.3 |
| fixed, nstrad = 3 (fluxes computed every 3rd step) | 99.1 |
| fixed, nstrad = 1 | 105.4 |

At T31 the shortwave is a small share of the step, so the saving is ~6%; it grows with resolution.

### 3c. Climatology, realistic Earth boundary conditions, 3 years

JCM alone at T31/L8 with the real T31 orography and land-sea mask and the climatological SST / sea-ice / land forcing shipped with jcm (`~/.cache/jcm/terrain_t31.nc`, `forcing_t31.nc`), 3 years from 2000-01-01, monthly means, last 2 years averaged; `nstrad` = 3 in both runs. The run and plotting scripts, the 1-year closure analysis and the timing script are available on request; all runs on a single V100.

| quantity (global, cos-lat weighted, last 2 yr) | shipped | fixed | fixed − shipped |
|---|---:|---:|---:|
| TOA net (W/m²) | 15.8 | 7.2 | −8.5 |
| surface net (W/m²) | −29.0 | 2.8 | +31.8 |
| TOA − surface imbalance (W/m²) | **44.8** | **4.5** | −40.3 |
| lowest-level air temperature (K) | 284.3 | 285.3 | +1.0 |
| mean atmospheric temperature (K) | 236.6 | 242.2 | +5.6 |
| precipitation (mm/day) | 3.48 | 2.75 | −0.72 |
| wall-clock per simulated year (s, V100) | 38.4 | 34.3 | −11% |

![zonal-mean T and u](https://raw.githubusercontent.com/dhruvbalwada/jax-gcm/docs/shortwave-issue/docs/issues/shortwave-cache/fig_zonal_T_u.png)
![surface air temperature and precipitation](https://raw.githubusercontent.com/dhruvbalwada/jax-gcm/docs/shortwave-issue/docs/issues/shortwave-cache/fig_surface.png)
![zonal energy budget](https://raw.githubusercontent.com/dhruvbalwada/jax-gcm/docs/shortwave-issue/docs/issues/shortwave-cache/fig_energy.png)
![global-mean time series](https://raw.githubusercontent.com/dhruvbalwada/jax-gcm/docs/shortwave-issue/docs/issues/shortwave-cache/fig_global_means.png)

What changes with the fix:

* The whole column warms, by a few K near the surface and up to ~20 K around the tropopause and in the stratosphere, where the ozone and water-vapour shortwave absorption that was being discarded is largest. Land surfaces warm by 1-4 K, oceans (prescribed SST) barely.
* Global precipitation drops by 25%, from 3.48 to 2.75 mm/day. With a third of the shortwave heating the column can only balance its longwave cooling by extra latent heat release, so the shipped model rains too much; 2.75 mm/day is close to the observed global mean (~2.7-2.9).
* The subtropical jets shift by a few m/s; the tropical column imbalance drops from ~100 W/m² to the level expected from residual spin-up.
* The 4.5 W/m² residual imbalance in the fixed run is spin-up drift (3 years from an isothermal start; both global-mean series are still trending), not a further bug.
* Throughput improves by ~11% end-to-end in this configuration because the shortwave and cloud routines are now genuinely skipped on two steps in three.

It may be worth checking whether the surface-temperature and jet biases relative to Fortran SPEEDY reported in the v1.1 paper (1.8 °C RMS, mostly over land) change with this fix, since v1.1.1 carries the bug.

## 4. Provenance

Full commit/PR trail with diff hunks: [HISTORY.md](https://raw.githubusercontent.com/dhruvbalwada/jax-gcm/docs/shortwave-issue/docs/issues/shortwave-cache/HISTORY.md). In short: PR #151 introduced the gate; an earlier attempt (PR #74, Oct 2024) was reverted before merge; PRs #177 and #196 touched the step alignment and the diagnostic persistence but not the tendency; no open issue reports this.

---
Found while coupling JCM in jax-esm, where the atmosphere's surface flux and the ocean's received flux could be compared directly. Analysis and fix prepared with the help of Claude (Anthropic); all numbers above were re-derived from runs on our machine.
